src/org/rascalmpl/library/ParseTree.rsc

```diff
@@ -811,3 +813,166 @@ bool isNonTerminalType(Symbol::\parameterized-sort(str _, list[Symbol] _)) = tru
+originally had a position will lose it unless ((reposition)) is configured to recompute it.
+
+By default ((reposition)) simulates the behavior of a ((parser)) exactly. Reparsing the 
+yield of a tree should always produce the exact same locations as ((reposition)) does.
+}
+@benefits{
+* Unlike reparsing, ((reposition)) will maintain all other keyword parameters of ((Tree)) nodes, like resolved qualified names and type attributes.
+* Can be used to erase superfluous annotations for memory efficiency, while keeping the essential ones.
+* 
+}
+&T <: Tree reposition(
+  &T <: Tree tree, 
+  loc file = tree@\loc.top, 
+  bool \markSyntax = true,
+  bool \markLexical = true,
+  bool \markSubLexical = false, 
+  bool \markRegular = true,
+  bool \markLayout = true,
+  bool \markSubLayout = true,
+  bool \markLit = false,
+  bool \markSubLit = false,
+  bool \markAmb = false,
+  bool \markCycle = false,
+  bool \markChar = false
+  ) {
+    // the cur variables are shared state by the `rec` local function that recurses over the entire tree
+    int curOffset = 0;
+    int curLine = 1;
+    int curColumn = 0;
+
+    @synopsis{Check if this rule is configured to be annotated}
+    default bool doAnno(Production _)               = false;
+    bool doAnno(prod(\lex(_), _, _))                = markLexical;
+    bool doAnno(prod(\label(_, \lex(_)), _, _))     = markLexical;
+    bool doAnno(prod(\layouts(_), _, _))            = markLayout;
+    bool doAnno(prod(\label(_, \layouts(_)), _, _)) = markLayout;
+    bool doAnno(prod(\sort(_), _, _))               = markSyntax;
+    bool doAnno(prod(\label(_, \sort(_)), _, _))    = markSyntax;
+    bool doAnno(\regular(_))                        = markRegular;
+    bool doAnno(prod(\lit(_), _, _))                = markLit;
+    bool doAnno(prod(\cilit(_), _, _))              = markLit;
+
+    @synopsis{Check if sub-structure of this rule is configured to be annotated}
+    default bool doSub(Production _)               = true;
+    bool doSub(prod(\lex(_), _, _))                = \markSubLexical;
+    bool doSub(prod(\label(_, lex(_)), _, _))      = \markSubLexical;
+    bool doSub(prod(\layouts(_), _, _))            = \markSubLayout;
+    bool doSub(prod(\label(_, \layouts(_)), _, _)) = \markSubLayout;
+    bool doSub(prod(\lit(_), _, _))                = \markSubLit;
+    bool doSub(prod(\cilit(_), _, _))              = \markSubLit;
+
+    // the character nodes drive the actual current position: offset, line and column
+    Tree rec(Tree t:char(int ch), bool _sub) {
+      beginOffset = curOffset;
+      beginLine   = curLine;
+      beginColumn = curColumn;
+
+      curOffset += 1;
+
+      switch (t) {
+        case ReturnChar _: {
+          curColumn = 0;
+        }
+      
+        case NewLineChar _ : {
+          curLine += 1;
+          curColumn = 0;
+        }
+      }
+
+      return markChar 
+        ? char(ch)[@\loc=file(beginOffset, 1, <beginLine, beginColumn>, <curLine, curColumn>)]
```

Comment:
With a Rascal extension based on Rascal `0.41.0-RC54`, I get a type error (that I do not understand) on this line:
> Invalid type: expected node, ADT, or concrete syntax types, found `![]`

Reply:
Me too. I think this was fixed in the meantime in the type-checker.